### PR TITLE
feat(parachain/availability-store): Confirm fork-awareness works in availability store subsystem

### LINKED ISSUE #3854: Confirm fork-awareness works in Availability Store Subsystem
## Issue summary

[The Polkadot Parachain Host Implementers' Guide](https://paritytech.github.io/polkadot-sdk/book/node/utility/availability-store.html#basic-scenarios-to-test) defines what this subsystem should do, including being fork-aware, which is when there are multiple candidate instances proposed for a given block, when that block is finalized it must be handled, and the other candidates removed.  The guide proposed the following test method:

- Fork-awareness of the relay chain is taken into account
  - Block B1 is added to the store.
  - Block B2 is added to the store.
  - Notify the subsystem that both B1 and B2 were included in different leafs of relay chain.
  - Notify the subsystem that the leaf with B1 was finalized.
  - Leaf with B2 is never finalized.
  - Leaf with B2 is pruned and its data is no longer available.
  - Wait until the finalized data of B1 should have been pruned.
  - B1 is no longer available.

## Other information and links

-

### PR DESCRIPTION
**Pull request by edwardmack**
## Changes

Adds tests to confirm fork-awareness works for availability store subsystem by testing the following:

- Block B1 is added to the store.
- Block B2 is added to the store.
- Notify the subsystem that both B1 and B2 were included in different leafs of relay chain.
- Notify the subsystem that the leaf with B1 was finalized.
- Leaf with B2 is never finalized.
- Leaf with B2 is pruned and its data is no longer available.
- Wait until the finalized data of B1 should have been pruned.
- B1 is no longer available.

## Tests

```sh
go test github.com/ChainSafe/gossamer/dot/parachain/availability-store -run ^TestForkfullnessWorks$ -v
```

## Issues

closes #3854 
## Primary Reviewer

@kishansagathiya

### CODE REVIEW

File: dot/parachain/availability-store/availability_store_test.go
Comment (jimjbrettj) [inline dot/parachain/availability-store/availability_store_test.go]: why did we need nolint here?

File: dot/parachain/availability-store/availability_store_test.go
Comment (jimjbrettj) [inline dot/parachain/availability-store/availability_store_test.go]: is nValidators number of validators?

File: dot/parachain/availability-store/availability_store_test.go
Comment (jimjbrettj) [inline dot/parachain/availability-store/availability_store_test.go:1937]: does the timing of this consistently work? A little curious about fitting in checks between sleeps, but if it always works that's alright

File: dot/parachain/availability-store/availability_store_test.go
Comment (jimjbrettj) [reply dot/parachain/availability-store/availability_store_test.go]: nit/perhaps numValidators or something

File: dot/parachain/availability-store/availability_store_test.go
Comment (edwardmack) [reply dot/parachain/availability-store/availability_store_test.go]: the linter complains 
``` 
`(*testHarness).hasAllChunks` - `nValidators` always receives `nValidators` (`10`) (unparam)
```
since nValidators is always 10, but it won't be in the real world, so this will change.

File: dot/parachain/availability-store/availability_store_test.go
Comment (edwardmack) [reply dot/parachain/availability-store/availability_store_test.go]: good idea, updated.

File: dot/parachain/availability-store/availability_store_test.go
Comment (edwardmack) [reply dot/parachain/availability-store/availability_store_test.go:1937]: It is working constantly, perhaps it's because it's sort of long, i suspect that if I reduced all the timing by a factor of ten (i.e. time.Sleep(300 * time.millisecond)) that it may cause issues, I'm not sure.

**Review by jimjbrettj — APPROVED** [review]
lgtm!

File: dot/parachain/availability-store/availability_store_test.go
Comment (kishansagathiya) [inline dot/parachain/availability-store/availability_store_test.go]: This function reduces redundancy. But CandidateReceipt has lot more things that pov hash. So, this function doesn't make much sense to me.
Might as well, use the code directly and let there be redundancy.

File: dot/parachain/availability-store/availability_store_test.go
Comment (kishansagathiya) [inline dot/parachain/availability-store/availability_store_test.go]: may be rename aLeaf to activatedLeaf?

File: dot/parachain/availability-store/availability_store_test.go
Comment (kishansagathiya) [inline dot/parachain/availability-store/availability_store_test.go]: lets use activatedLeaf instead of aLeaf?

File: dot/parachain/availability-store/availability_store_test.go
Comment (kishansagathiya) [inline dot/parachain/availability-store/availability_store_test.go:1880]: What are the chances that this goroutine keeps running even if this test finishes?

**Review by kishansagathiya — COMMENTED** [review]
Had a call with Ed, he walked me through the code to review it.
LGTM
Will approve, after base branch is merged.

File: dot/parachain/availability-store/availability_store_test.go
Comment (edwardmack) [reply dot/parachain/availability-store/availability_store_test.go]: Good point, I've updated this.

File: dot/parachain/availability-store/availability_store_test.go
Comment (edwardmack) [reply dot/parachain/availability-store/availability_store_test.go]: Updated.

File: dot/parachain/availability-store/availability_store_test.go
Comment (edwardmack) [reply dot/parachain/availability-store/availability_store_test.go]: Updated.

File: dot/parachain/availability-store/availability_store_test.go
Comment (edwardmack) [reply dot/parachain/availability-store/availability_store_test.go:1880]: I've added defer stop to these tests to handle this.

**Review by kishansagathiya — APPROVED** [review]

**Pull request merged**
